### PR TITLE
support 2.7 as well

### PR DESCRIPTION
**Pull request by itdependsnetworks**

### CODE REVIEW

**Comment by jtdub** [comment]
Hey @itdependsnetworks - I'm going to discuss this with the team. Our current tooling is ~python 3.6 only. There is a possibility that we will be implementing python 3+ code objects, such as type hinting (https://www.python.org/dev/peps/pep-0484/).

File: hier_config/hc_child.py [hunk 1]
```diff
@@ -569,7 +572,7 @@ def _config_to_get_to_left(self, target, delta):
                 deleted.negate()
                 if self_child.children:
                     deleted.comments.add(
-                        f"removes {len(self_child.children_dict) + 1} lines")
+                        "removes {} lines".format(len(self_child.children_dict)))
```
Comment (aedwardstx) [inline hier_config/hc_child.py]: The `len(self_child.children_dict) + 1` counts the negation of the child and all of its children.  We will want to keep the `+ 1`.

**Comment by itdependsnetworks** [comment]
@aedwardstx thanks! I did not mean to do that.

@jtdub I believe there is a (sub-optimal) way to accommodate both, and while it would be great if everything was 3.X in the world, we are where we are. IMHO this would broaden the scope of the project.

**Comment by jtdub** [comment]
Just checked Travis, the unittests failed, but it's not reporting properly:

```
$ python ./tests/tests.py
.......................F..............................
======================================================================
FAIL: test_dump_and_load_from_dump_and_compare (test_hier_config.TestHConfig)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/build/netdevops/hier_config/tests/test_hier_config.py", line 90, in test_dump_and_load_from_dump_and_compare
    self.assertEqual(hier_pre_dump, hier_post_dump)
AssertionError: <instance object at 0x7f7605bb1908> != <instance object at 0x7f7605bb1b48>
----------------------------------------------------------------------
Ran 54 tests in 0.038s
FAILED (failures=1)
```
I'll look into that failure to catch the failure when I get a chance.

File: .travis.yml [hunk 2]
```diff
@@ -2,6 +2,7 @@
 language: python
 
 python:
+- "2.7"
 - "3.6"
```
Comment (jtdub) [inline .travis.yml]: By pulling out the `f` string and `yield from`, this should make it compatible with anything above 2.7. It'd be a good idea to add support for 3.0 - 3.5 as well.

File: setup.py [hunk 3]
```diff
@@ -15,7 +15,7 @@
     author="Andrew Edwards, Jan Brooks, James Williams",
     author_email="andrew.edwards@rackspace.com, jan.brooks@rackspace.com, james.williams@rackspace.com",
     keywords = "hier_config",
-    python_requires='~=3.6',
```
Comment (jtdub) [inline setup.py]: I believe this should be changed to `>=2.7`

File: setup.py [hunk 4]
```diff
@@ -24,6 +24,7 @@
         'Intended Audience :: Telecommunications Industry',
         'License :: OSI Approved :: MIT License',
         'Programming Language :: Python :: 3.6',
```
Comment (jtdub) [inline setup.py]: We could call out the in between python versions as well.

**Comment by itdependsnetworks** [comment]
Yay!! @jmcgill298

File: hier_config/hc_child.py [hunk 5]
```diff
@@ -383,16 +384,18 @@ def all_children_sorted_by_tags(self, include_tags, exclude_tags):
...
+            for child in children.all_children_sorted():
+                yield child
 
     def all_children(self):
         """ Recursively find and yield all children """
 
-        for child in self.children:
-            yield child
-            yield from child.all_children()
+        for children in self.children:
```
Comment (aedwardstx) [inline hier_config/hc_child.py]: This should be `for child in self.children:`

File: hier_config/hc_child.py [hunk 6]
```diff
@@ -383,16 +384,18 @@ def all_children_sorted_by_tags(self, include_tags, exclude_tags):
...
+            yield children
+            for child in children.all_children():
```
Comment (aedwardstx) [inline hier_config/hc_child.py]: Change this to:
```
for sub_child in child.all_children():
    yield sub_child
```

File: hier_config/hc_child.py [hunk 7]
```diff
@@ -383,16 +384,18 @@ def all_children_sorted_by_tags(self, include_tags, exclude_tags):
     def all_children_sorted(self):
         """ Recursively find and yield all children sorted at each hierarchy """
 
-        for child in sorted(self.children):
-            yield child
-            yield from child.all_children_sorted()
+        for children in sorted(self.children):
```
Comment (aedwardstx) [inline hier_config/hc_child.py]: This should be:
```
for child in sorted(self.children):
    yield child
    for sub_child in child.all_children_sorted():
        yield sub_child
```

**Comment by aedwardstx** [comment]
I would like to implement Type Hints at some point. Doing so does not seem compatible with Python2.7.  I don't suppose its possible for the customer's application to be upgraded to Python3? :-) If we do add Python2.7 support, I think we should restrict it to 2.7 and 3.6 to make it more practical to deprecate Python2.7 at some point to support Type Hints.

**Comment by jmcgill298** [comment]
Have you tried using comment-based syntax? 
http://mypy.readthedocs.io/en/latest/python2.html

File: setup.py (hunk 3 above)
Comment (itdependsnetworks) [reply setup.py]: This has been updated, but I guess if you comment on the delete it doesn't' reflect.

**Comment by aedwardstx** [comment]
I'm good with supporting Python2.7. Comment based syntax is an option but it wouldn't allow for using packages like https://github.com/RussBaz/enforce. Once Python2.7 is retired, I vote we deprecate support for it https://pythonclock.org/.  

Another option to consider:
We've realized that once you're building remediations for hundreds or thousands of devices at a time and have pretty complex remediation fix up functions, it becomes handy to front the remediation build logic with an API or worker queue model. This has the added benefit of allowing HConfig to run in another runtime.

File: tests/test_hier_config.py [hunk 8]
```diff
@@ -86,8 +86,16 @@ def test_dump_and_load_from_dump_and_compare(self):
 
         hier_post_dump = HConfig(self.host_a, self.os, self.options)
         hier_post_dump.load_from_dump(dump)
+        # py2.7 does not eval an object the same way
```
Comment (aedwardstx) [inline tests/test_hier_config.py]: What do you mean by this? https://github.com/netdevops/hier_config/blob/master/hier_config/hc_child.py#L70 Implements a custom __eq__ method to handle this case specifically.

Comment (jmcgill298) [reply tests/test_hier_config.py]: https://github.com/netdevops/hier_config/blob/ba00aa2063c6883544679b652a28e687d21226bf/hier_config/hc_child.py#L88

I believe the issue is that you only implemented a custom __eq__ method and not the other comparison methods, particularly __ne__ for this example.

Comment (jmcgill298) [reply tests/test_hier_config.py]: perhaps a  quick fix would be `if not self_child == other_child:`

Comment (aedwardstx) [reply tests/test_hier_config.py]: Per https://stackoverflow.com/questions/4352244/python-should-i-implement-ne-operator-based-on-eq Python3 automatically understands the relationship between `==` and `!=`. Sounds like the following should be added if we wish to facilitate Python2.7 support:
```Python
def __ne__(self, other):
    return not self.__eq__(other)
```

File: hier_config/__init__.py [hunk 9]
```diff
@@ -90,9 +90,12 @@ def __eq__(self, other):
         for self_child, other_child in zip(sorted(self.children), sorted(other.children)):
             if self_child != other_child:
                 return False
-
         return True
 
+    def __ne__(self, other):
```
Comment (aedwardstx) [inline hier_config/__init__.py]: Since this class inherits `HConfigChild` and `__ne__` is defined there, is there a need to define `__ne__` here?

**Comment by jtdub** [comment]
Can you increment `hier_config.__version__` to `1.2.0`?

**Comment by aedwardstx** [comment]
Looks like this needs to be rebased before merging.

**Pull request merged**